### ISSUE TITLE
Add Node::pauseSelfAndChildren, Node::resumeSelfAndChildren

### ISSUE THREAD

**Issue by pandamicro** [cocos2d/cocos2d-x#13615]
Currently, Node::pause and Node::resume only affect the node itself. But there are some users request to add those APIs, because it will be easier to control the whole sub node tree.

Here is the code proposed by user

```
void Node::resume()
{
    _scheduler->resumeTarget(this);
    _actionManager->resumeTarget(this);
    _eventDispatcher->resumeEventListenersForTarget(this);

    Vector<Node*> ss = this->getChildren();

    for (Vector<Node*>::iterator iter = ss.begin();iter != ss.end();iter++) {
        auto n = (Node*)*iter;
        n->resume();
    }
}

void Node::pause()
{
    _scheduler->pauseTarget(this);
    _actionManager->pauseTarget(this);
    _eventDispatcher->pauseEventListenersForTarget(this);

    Vector<Node*> ss = this->getChildren();

    for (Vector<Node*>::iterator iter = ss.begin();iter != ss.end();iter++) {
        auto n = (Node*)*iter;
        n->pause();
    }
}
```

@minggo @zilongshanren Thoughts ?

**Comment by milos1290** [comment]
This should be implemented. But this can have some negative effects. For example if you expect child node to receive events from EventDispatcher, that wont work anymore since that child node will be paused too. I had a problem with this issue i was pausing the whole game scene like this (without HUD), and if user went to background on Android and returned back to game, my custom mesh couldn't be recreated because EventDispatcher for that particular node was paused, so i had to manually resume that node so that i could receive EVENT_RENDERER_RECREATED event.

**Comment by dumganhar** [comment]
Is it better to add an argument `pauseChildren` for that purpose? Something like:

``` c++
void pause(bool toPauseChildren = false)
void Node::pause(bool toPauseChildren)
{
    _scheduler->pauseTarget(this);
    _actionManager->pauseTarget(this);
    _eventDispatcher->pauseEventListenersForTarget(this);

    if (toPauseChildren) {
    Vector<Node*> ss = this->getChildren();

    for (Vector<Node*>::iterator iter = ss.begin();iter != ss.end();iter++) {
        auto n = (Node*)*iter;
        n->pause();
    }
    }
}
```

And it's default value is false for compatibility.

**Comment by pandamicro** [comment]
@dumganhar I personally like more intuitive APIs with clear names to identify their usage, although it's true that we have many APIs similar to the one you suggested.

@milos1290 I see the problem, it's not really relative to this issue, but it indeed could be improved. I think the problem is that, paused state should still receive some events, I don't know how, but it means we should introduce a new mechanism into Event Dispatcher

**Comment by milos1290** [comment]
Yeah of course, but still pausing the node won't receive that callback. Maybe we should leave EventDispatcher active always, and user can check if node is paused before reacting on received event. This way engine implementation of nodes specific events will be active always, and user can choose to react to their own events based on node state.

**Comment by mertkasar** [comment]
Any news on that improvement? Because I think that this one is essential. Any dirty flag like system that I worked with in past operates in this way.

But with cocos2d-x, there are tons of overrided pause and resume methods all over my code base.
